### PR TITLE
Intl 2021 Updates

### PR DESCRIPTION
**Pull request by orta**
Re: #45646 - no breaking changes in here though

This is quite a few PRs + issues combined, rebased etc:

 - #39664
 - #45420
 - #44022 
 - #42944
 - #45652

I've tried to reduce the overall character count for some of the docs (e.g. removing some JSDocs which are redundant because the type system will tell you the types of things) and removed all the specs links, I can't imagine your average `Intl` API consumer wanting to read the spec (over MDN).

Also adds some fresh docs about how lib files work to the README, covering how `Intl` is a wee bit special in comparison to some of the other cases we have in the `.d.ts` files.

### CODE REVIEW

**Comment by orta** [comment]
@typescript-bot run dt

**Comment by orta** [comment]
I've double checked all the es5 `Intl` properties, and they are in the right place (they were [spec'd in 2012](https://402.ecma-international.org/1.0/) which is a reasonable approximation IMO)

**Comment by orta** [comment]
@longlho any chance I could get your eyes over this?

**Comment by longlho** [comment]
Yup will take a look asap

File: src/lib/es2020.intl.d.ts [hunk 1]
```diff
@@ -262,28 +198,154 @@ declare namespace Intl {
...
+    }
+
+    interface DateTimeFormatOptions {
+        calendar?: string | undefined;
+        dayPeriod?: "narrow" | "short" | "long" | undefined;
+        numberingSystem?: string | undefined;
+        hourCycle?: "h11" | "h12" | "h23" | "h24" | undefined;
     }
 
     interface DateTimeFormatOptions {
```
Comment (sandersn) [inline src/lib/es2020.intl.d.ts:223]: Probably should merge these two DateTimeFormatOptions declarations

File: src/lib/es2020.intl.d.ts [hunk 2]
```diff
@@ -262,28 +198,154 @@ declare namespace Intl {
...
+        timeStyle?: "full" | "long" | "medium" | "short" | undefined;
+        hourCycle?: "h11" | "h12" | "h23" | "h24" | undefined;
+    }
+
+    type LocaleHourCycleKey = "h12" | "h23" | "h11" | "h24";
+    type LocaleCollationCaseFirst = "upper" | "lower" | "false";
+
+    interface LocaleOptions {
+        /** A string containing the language, and the script and region if available. */
+        baseName?: string;
```
Comment (sandersn) [inline src/lib/es2020.intl.d.ts:234]: shouldn't all these optional properties add `| undefined`?

**Review by sandersn — CHANGES_REQUESTED** [review]
Just a lot of minor edit suggestions.

File: src/lib/es2020.intl.d.ts (hunk 2 above)
Comment (orta) [reply src/lib/es2020.intl.d.ts:234]: I added the `| undefined` to the user inputs, but not the returned objects from the API - so _I think_ that leaves this as OK but will re-check

File: src/lib/es2018.intl.d.ts [hunk 3]
```diff
@@ -33,9 +33,16 @@ declare namespace Intl {
     const PluralRules: {
         new (locales?: string | string[], options?: PluralRulesOptions): PluralRules;
         (locales?: string | string[], options?: PluralRulesOptions): PluralRules;
-        supportedLocalesOf(
-            locales: string | string[],
-            options?: PluralRulesOptions,
-        ): string[];
+
+        supportedLocalesOf(locales: string | string[], options?: { localeMatcher?: "lookup" | "best fit" }): string[];
```
[the remark this answers is not in the document]
Comment (longlho) [reply src/lib/es2018.intl.d.ts:37]: yeah this seems correct per https://402.ecma-international.org/5.0/#sec-supportedlocales which is what `supportedLocalesOf` calls

File: src/lib/es2018.intl.d.ts [hunk 4]
```diff
@@ -33,9 +33,16 @@ declare namespace Intl {
...
-        supportedLocalesOf(
-            locales: string | string[],
-            options?: PluralRulesOptions,
-        ): string[];
+
+        supportedLocalesOf(locales: string | string[], options?: { localeMatcher?: "lookup" | "best fit" }): string[];
     };
+
+    interface NumberFormatPart {
+        type: "compact" | "currency" | "decimal" | "exponentInteger" | "exponentMinusSign" | "exponentSeparator" | "fraction" | "group" | "infinity" | "integer" | "literal" | "minusSign" | "nan" | "plusSign" | "percentSign" | "unit" | "unknown";
```
Comment (longlho) [inline src/lib/es2018.intl.d.ts:41]: This seems to be the es2020 set, es2018 parts don't have `compact | exponentInteger | exponentMinusSign | exponentSeparator | unit | unknown`. 

- `compact` comes from compact notation which is es2020
- `exponentInteger | exponentMinusSign | exponentSeparator` comes from scientific notation in es2020
- `unit` comes from unit formatting in es2020
- `unknown` is just `literal`.

This is missing `symbol` from `currency` formatting as well.

Overall I think the set for es2018 is `literal | nan | infinity | percent | integer | group | decimal | fraction | plusSign | minusSign | percentSign | currency | code | symbol | name`

Reference: https://402.ecma-international.org/5.0/#sec-partitionnumberpattern

File: src/lib/es2021.intl.d.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+declare namespace Intl {
+
+    interface DateTimeFormatOptions {
```
Comment (longlho) [inline src/lib/es2021.intl.d.ts:3]: this is missing `formatMatcher` I believe

File: src/lib/es2021.intl.d.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+declare namespace Intl {
```
Comment (longlho) [inline src/lib/es2021.intl.d.ts:1]: This is missing `formatRange` & `formatRangeToParts` from `Intl.DateTimeFormat`
Ref: https://tc39.es/ecma402/2021/#sec-intl.datetimeformat.prototype.formatRange

**Review by longlho — CHANGES_REQUESTED** [review]
Overall I think es2021 is missing a ton of stuff but I think it's a good enough set of changes in the scaffolding for others to fill in.

File: src/lib/es2018.intl.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (orta) [reply src/lib/es2018.intl.d.ts:37]: Yeah, I think this is the right thing for now until this gets big enough to warrant its own type - don't have strong opinions on it though

**Comment by orta** [comment]
OK, thanks for the feedback everyone - assuming this is good I will merge weds and start closing all the linked PRs

File: src/lib/es2021.intl.d.ts [hunk 7]
```diff
@@ -1,9 +1,15 @@
 declare namespace Intl {
 
     interface DateTimeFormatOptions {
+        formatMatcher?: "basic" | "best fit" | "best fit";
```
Comment (longlho) [inline src/lib/es2021.intl.d.ts:4]: do you wanna add `| undefined` as well?

File: src/lib/es2020.intl.d.ts [hunk 8]
```diff
@@ -197,6 +197,10 @@ declare namespace Intl {
         ): UnicodeBCP47LocaleIdentifier[];
     };
 
+    interface NumberFormatPart {
+        type: "compact" | "currency" | "decimal" | "exponentInteger" | "exponentMinusSign" | "exponentSeparator" | "fraction" | "group" | "infinity" | "integer" | "literal" | "minusSign" | "nan" | "plusSign" | "percentSign" | "unit" | "unknown";
```
Comment (longlho) [inline src/lib/es2020.intl.d.ts:201]: I'd restructure this to something like
```ts
// ES2018
type ES2018NumberFormatPartType = "literal" | "nan" | "infinity" | "percent" | "integer" | "group" | "decimal" | "fraction" | "plusSign" | "minusSign" | "percentSign" | "currency" | "code" | "symbol" | "name"
// ES2020
type ES2020NumberFormatPartType = ES2018NumberFormatPartType | "compact" | "exponentInteger" | "exponentMinusSign" | "exponentSeparator" | "unit" | "unknown"

interface NumberFormatPart {
	type: ES2020NumberFormatPartType
	value: string
}
```

Or at least keep the order the same across ES versions so it's easier to review

**Review by longlho — APPROVED** [review]

**Comment by longlho** [comment]
Editorial suggestions but overall lgtm

File: src/lib/es2020.intl.d.ts (hunk 8 above)
Comment (orta) [reply src/lib/es2020.intl.d.ts:201]: Great ideas, implemented 👍🏻  and pushed

**Comment by oceandrama** [comment]
@orta
Hi! Could you add typings for `DateTimeFormat.formatRange` (https://github.com/tc39/proposal-intl-DateTimeFormat-formatRange)?
This is already in stage 4, but I couldn't find the types for it in this repository

**Comment by longlho** [comment]
@oceandrama I think that can be a follow-up diff. ES2021 in this diff is still missing a bunch of normative changes

**Review by sandersn — APPROVED** [review]

**Comment by orta** [comment]
Thanks folks

**Pull request merged**